File: src/main/java/io/github/slimefunguguproject/bump/core/handlers/BowUseHandler.java
package io.github.slimefunguguproject.bump.core.handlers;

import java.util.Optional;

import javax.annotation.Nonnull;

import org.bukkit.Material;
import org.bukkit.entity.Player;
import org.bukkit.event.entity.EntityShootBowEvent;
import org.bukkit.inventory.ItemStack;

import io.github.slimefunguguproject.bump.implementation.listeners.BowShootListener;
import io.github.thebusybiscuit.slimefun4.api.exceptions.IncompatibleItemHandlerException;
import io.github.thebusybiscuit.slimefun4.api.items.ItemHandler;
import io.github.thebusybiscuit.slimefun4.api.items.SlimefunItem;

/**
 * This {@link ItemHandler} is triggered when a player right click with a bow.
 *
 * @author ybw0014
 * @see BowShootListener
 */
public interface BowUseHandler extends ItemHandler {

    void onUse(EntityShootBowEvent e, Player p, ItemStack bow);

    @Nonnull
    @Override
    default Optional<IncompatibleItemHandlerException> validate(SlimefunItem item) {
        if (item.getItem().getType() != Material.BOW) {
            return Optional.of(new IncompatibleItemHandlerException("Only bows can have a BowUseHandler.", item, this));
        }

        return Optional.empty();
    }

    @Nonnull
    @Override
    default Class<? extends ItemHandler> getIdentifier() {
        return BowUseHandler.class;
    }
}

File: src/main/kotlin/io/github/slimefunguguproject/bump/core/handlers/BowUseHandler.kt
package io.github.slimefunguguproject.bump.core.handlers

import io.github.thebusybiscuit.slimefun4.api.exceptions.IncompatibleItemHandlerException
import io.github.thebusybiscuit.slimefun4.api.items.ItemHandler
import io.github.thebusybiscuit.slimefun4.api.items.SlimefunItem
import org.bukkit.Material
import org.bukkit.entity.Player
import org.bukkit.event.entity.EntityShootBowEvent
import org.bukkit.inventory.ItemStack
import java.util.Optional

fun interface BowUseHandler : ItemHandler {
    fun onUse(e: EntityShootBowEvent, p: Player, bow: ItemStack)

    override fun validate(item: SlimefunItem): Optional<IncompatibleItemHandlerException> {
        if (item.item.type != Material.BOW) {
            return Optional.of(IncompatibleItemHandlerException("Only bows can have a BowUseHandler.", item, this))
        }

        return Optional.empty()
    }

    override fun getIdentifier() = BowUseHandler::class.java
}

